spec/std/hash_spec.cr

```diff
@@ -1393,16 +1393,25 @@ describe "Hash" do
     hash.@indices_size_pow2.should eq(12)
   end
 
-  it "rehashes" do
-    a = [1]
-    h = {a => 0}
-    (10..100).each do |i|
-      h[[i]] = i
-    end
-    a << 2
-    h[a]?.should be_nil
-    h.rehash
-    h[a].should eq(0)
+  describe "#rehash" do
+    it "rehashes" do
+      a = [1]
+      h = {a => 0}
+      (10..100).each do |i|
+        h[[i]] = i
+      end
+      a << 2
+      h[a]?.should be_nil
+      h.rehash
+      h[a].should eq(0)
+    end
+
+    it "resets @first (#14602)" do
+      h = {"a" => 1, "b" => 2}
+      h.delete("a")
+      h.rehash
+      h["b"].should eq 2
```

Comment:
I'd suggest testing the hash itself to avoid unintended side-effects.

```suggestion
      h.should eq({"b" => 2})
```

Reply:
That won't do because equality holds even without this patch.
`Hash#==` checks if the sizes are identical (which they are because `size` is independent from `@first`) and then iterates all entries in `self` to see if they are in `other`. Since iterating `self` does not yield, this check succeeds.
I suppose we could do something else that involves `#each`, like test the result of `#to_a` (although that causes invalid memory access without this patch).

Reply:
I changed the expectation to use `#to_s`.